scalardb/src/scalardb/transfer.clj

```diff
@@ -125,37 +125,39 @@
         (assoc op :type :fail :error {:results results})))))
 
 (defn- read-record
-  "Read a record with a transaction. If read fails, an exception is thrown."
-  [tx storage i]
-  ;; Need Storage API to read the transaction metadata
-  (let [tx-result (.get tx (prepare-get i))
-        result (.get storage (prepare-get i))]
+  "Read and update the specified record with a transaction"
+  [test id]
+  (let [tx (scalar/start-transaction test)
+        tx-result (.get tx (prepare-get id))
+        ;; Need Storage API to read the transaction metadata
+        result (.get @(:storage test) (prepare-get id))]
     ;; Put the same balance to check conflicts with in-flight transactions
     (->> (calc-new-balance tx-result 0)
-         (prepare-put i)
+         (prepare-put id)
          (.put tx))
+    (.commit tx)
     result))
 
-(defn read-all-with-retry
-  [test n]
-  (scalar/check-transaction-connection! test)
-  (scalar/check-storage-connection! test)
+(defn- read-record-with-retry
+  [test id]
   (scalar/with-retry
     (fn [test]
       (scalar/prepare-transaction-service! test)
       (scalar/prepare-storage-service! test))
     test
     (try
-      (let [tx (scalar/start-transaction test)
-            results (doall (map #(read-record tx @(:storage test) %)
-                                (range n)))]
-        (.commit tx)
-        results)
+      (read-record test id)
       (catch Exception e
```

Comment:
I think we can catch `CrudConflictException` if it's for the lazy recovery. What do you think?

Reply:
@brfrn169 Another exception could be thrown from the commit, couldn't it?

Reply:
@yito88 Sorry, it seems I had the wrong idea. Please ignore the comment.

So currently, we retry all the record reads, but after this PR, we will retry each record read individually. Is my understanding correct?

Reply:
@brfrn169 No problem. "read-record" tries to update, so it's a bit misleading.

> we retry all the record reads together, but after this PR, we will retry each record read individually. Is my understanding correct?

Exactly!

Reply:
Thanks!